pkg/plugins/golang/options.go

```diff
@@ -0,0 +1,256 @@
+	// NOTE: We must do this for all the required flags first or we may output the wrong
+	// error as flags may seem to be missing because Cobra assigned them to another flag.
+	if strings.HasPrefix(opts.Group, "-") {
+		return fmt.Errorf(groupPresent)
+	}
+	if strings.HasPrefix(opts.Version, "-") {
+		return fmt.Errorf(versionPresent)
+	}
+	if strings.HasPrefix(opts.Kind, "-") {
+		return fmt.Errorf(kindPresent)
+	}
+	// Now we can check that all the required flags are not empty
+	if len(opts.Version) == 0 {
+		return fmt.Errorf(versionRequired)
+	}
+	if len(opts.Kind) == 0 {
+		return fmt.Errorf(kindRequired)
+	}
+
+	// Check if the qualified group has a valid DNS1123 subdomain value
+	if err := validation.IsDNS1123Subdomain(opts.QualifiedGroup()); err != nil {
+		return fmt.Errorf("either group or domain is invalid: (%v)", err)
+	}
+
+	// Check if the version follows the valid pattern
+	if !versionRegex.MatchString(opts.Version) {
+		return fmt.Errorf("version must match %s (was %s)", versionPattern, opts.Version)
+	}
+
+	validationErrors := []string{}
+
+	// Require kind to start with an uppercase character
+	if string(opts.Kind[0]) == strings.ToLower(string(opts.Kind[0])) {
+		validationErrors = append(validationErrors, "kind must start with an uppercase character")
+	}
+
+	validationErrors = append(validationErrors, validation.IsDNS1035Label(strings.ToLower(opts.Kind))...)
+
+	if len(validationErrors) != 0 {
+		return fmt.Errorf("invalid Kind: %#v", validationErrors)
+	}
+
+	// TODO: validate plural strings if provided
+
+	// Ensure apiVersions for k8s types are empty or valid.
+	for typ, apiVersion := range map[string]string{
+		"CRD":     opts.CRDVersion,
+		"Webhook": opts.WebhookVersion,
+	} {
+		switch apiVersion {
+		case "", "v1", "v1beta1":
+		default:
+			return fmt.Errorf("%s version must be one of: v1, v1beta1", typ)
+		}
+	}
+
+	return nil
+}
+
+// QualifiedGroup returns the fully qualified group name with the available information.
+func (opts Options) QualifiedGroup() string {
+	switch "" {
+	case opts.Domain:
+		return opts.Group
+	case opts.Group:
+		return opts.Domain
+	default:
+		return fmt.Sprintf("%s.%s", opts.Group, opts.Domain)
+	}
+}
+
+// GVK returns the GVK identifier of a resource.
+func (opts Options) GVK() resource.GVK {
+	return resource.GVK{
+		Group:   opts.QualifiedGroup(),
+		Version: opts.Version,
+		Kind:    opts.Kind,
+	}
+}
+
+// NewResource creates a new resource from the options
+func (opts Options) NewResource(c *config.Config) *resource.Resource {
+	res := &resource.Resource{
+		Group:      opts.Group,
+		Domain:     opts.Domain,
+		Version:    opts.Version,
+		Kind:       opts.Kind,
+		Path:       resource.LocalPath(c.Repo, opts.Group, opts.Version, c.MultiGroup),
+		Controller: opts.DoController,
+	}
+
+	if opts.Plural != "" {
+		res.Plural = opts.Plural
+	} else {
+		// If not provided, compute a plural for for Kind
+		res.Plural = resource.RegularPlural(opts.Kind)
+	}
```

Comment:
```suggestion
	// If not provided, compute a plural for for Kind
	plural := opts.Plural
	if plural == "" {
		plural = flect.Pluralize(strings.ToLower(opts.Kind))
	}
```

Reply:
Your provided way requires an extra variable (allocation, garbage collection, ...), an `if` conditional, and at most 3 assign operations (`plural := opts.Plural`, `plural = flect.Pluralize(strings.ToLower(opts.Kind))`, and `res,Plural = plural`). Mine only requires an `if` conditional and at most 1 assign operation (either `res.Plural = opts.Plural` or `res.Plural = resource.RegularPlural(opts.Kind)`).

Reply:
Note that you felt compiled to add comments for all these if/else(s) to make it make sense.
So, IMO I'd like to keep the code as it was before (without all these else's and comments) which is more readable/understandable and in this way easy to keep maintained.

Reply:
The comment was already present in the previous code